File: crates/axe-core/src/context.rs
//! Thin `Context`: the three things every read-command needs after init.
//!
//! Commands that operate on an existing install load a `Context` instead of
//! repeating "parse config → probe layout → handle errors" themselves.

use camino::Utf8PathBuf;

use crate::error::CoreError;
use crate::layout::Layout;
use crate::{Config, probe};

/// Everything a read-command needs, resolved from one `axe.toml`.
#[derive(Debug, Clone)]
pub struct Context {
    pub config_path: Utf8PathBuf,
    pub config: Config,
    pub layout: Layout,
}

impl Context {
    /// Load `axe.toml`, probe the install root, return a fully-resolved context.
    ///
    /// # Errors
    /// Returns [`CoreError::Config`] if the config file cannot be read or parsed,
    /// or [`CoreError::Discovery`] if the install root cannot be probed.
    pub fn load(config_path: &Utf8PathBuf) -> Result<Self, CoreError> {
        let config = Config::load(config_path)?;
        let layout = probe::probe(&config.server.root)?;
        Ok(Self {
            config_path: config_path.clone(),
            config,
            layout,
        })
    }
}

File: src/core/context.ts
import { type Config, loadConfig } from './config.ts';
import type { Layout } from './layout.ts';
import { probe } from './probe.ts';

export type Context = {
  config: Config;
  layout: Layout;
};

/**
 * Load `axe.toml` and probe its `[server].root`, returning both as a typed
 * Context. Used by every command that operates on an existing install.
 *
 * Not a class. Not a god object. Two operations called together because
 * they're always called together. Throws `AxeError({kind:'config'})` on
 * config failure, `AxeError({kind:'discovery'})` on probe failure.
 */
export async function loadContext(configPath: string): Promise<Context> {
  const config = await loadConfig(configPath);
  const layout = await probe(config.server.root);
  return { config, layout };
}
